File: cs/Problems/Day10.cs
namespace aoc24.Problems;

// https://adventofcode.com/2024/day/10
public sealed class Day10 : IProblem<int>
{
    public int Solve(string input) => CountTrailHeadScoresOptimized(input);

    private readonly HashSet<(Point, Point)> paths = [];

    public int CountTrailHeadScoresOptimized(ReadOnlySpan<char> input)
    {
        Span<char> inner = stackalloc char[CharMatrix.SizeFor(input)];
        var matrix = CharMatrix.CreateFrom(input, inner);

        paths.Clear();

        for (int y = 0; y < matrix.Height; y++)
        {
            for (int x = 0; x < matrix.Width; x++)
            {
                var start = new Point(x, y);
                BuildTrailHeadPaths(matrix, start, start, paths);
            }
        }

        return paths.Count;
    }

    private static void BuildTrailHeadPaths(
        CharMatrix matrix,
        Point start,
        Point current,
        HashSet<(Point, Point)> paths,
        int expected = 0
    )
    {
        int value = matrix.CharAt(current) is char curr ? curr - '0' : int.MinValue;

        if (value != expected)
            return;

        if (value == expected && expected == 9)
        {
            paths.Add((start, current));
            return;
        }

        foreach (var direction in NavigableDirections)
            BuildTrailHeadPaths(matrix, start, current + Point.FromDirection(direction), paths, expected + 1);
    }

    private static readonly Direction[] NavigableDirections = [Direction.Right, Direction.Down, Direction.Left, Direction.Up];
}

File: go/problems/day10.go
package problems

import (
	"image"
	"math"
	"strings"

	"github.com/Meryovi/hobby-aoc-2024/go/shared"
)

type Day10 struct{}

type pointPair struct {
	start image.Point
	end   image.Point
}

func (d Day10) Solve(input string) int {
	paths := make(map[pointPair]pointPair)
	matrix := strings.Split(input, shared.NewLine)

	for y := range matrix {
		for x := range matrix[0] {
			start := image.Pt(x, y)
			buildTrailPaths(matrix, start, start, paths, 0)
		}
	}

	return len(paths)
}

var navDirs = []image.Point{right, down, left, up}

func buildTrailPaths(matrix []string, start, curr image.Point, paths map[pointPair]pointPair, expected int) {
	val := math.MinInt
	if curr.Y >= 0 && curr.X >= 0 && curr.Y < len(matrix) && curr.X < len(matrix[0]) {
		val = int(matrix[curr.Y][curr.X] - '0')
	}

	if val != expected {
		return
	}

	if expected == 9 {
		pp := pointPair{start, curr}
		paths[pp] = pp
		return
	}

	for _, dir := range navDirs {
		buildTrailPaths(matrix, start, curr.Add(dir), paths, expected+1)
	}
}
